website/oauth/models/__init__.py

```diff
@@ -348,3 +355,68 @@ def handle_callback(self, response):
         :return dict:
         """
         pass
+
+
+class ApiOAuth2Application(StoredObject):
+    """Registration and key for user-created OAuth API applications"""
+    _id = fields.StringField(
+        primary=True,
+        default=lambda: str(ObjectId())
+    )
+
+    # Client ID and secret. Use separate ID field so ID format doesn't have to be restricted to database internals.
+    client_id = fields.StringField(default=lambda: uuid.uuid4().hex,  # Not *guaranteed* unique, but very unlikely
+                                   unique=True,
+                                   index=True)
+    client_secret = fields.StringField(default=generate_client_secret)
+
+    active = fields.BooleanField(default=True,  # Set to False if application is deactivated
+                                 index=True)
+
+    owner = fields.ForeignField('User',
+                                backref='created',
+                                index=True,
+                                required=True)
+
+    # User-specified application descriptors
+    name = fields.StringField(index=True, required=True)
+    description = fields.StringField(required=False)
+
+    date_created = fields.DateTimeField(auto_now_add=datetime.datetime.utcnow,
+                                        editable=False)
+
+    home_url = fields.StringField(required=True,
+                                  validate=URLValidator())
+    callback_url = fields.StringField(required=True,
+                                      validate=URLValidator())
+
+    def deactivate(self):
+        """
+        Deactivate an ApiOAuth2Application
+
+        Does not delete the database record, but revokes all tokens and sets a flag that hides this instance from API
+        """
+        client = cas.get_client()
+        # Will raise a CasHttpError if deletion fails, which will also stop setting of active=False.
+        resp = client.revoke_application_tokens(self.client_id, self.client_secret)  # noqa
+
+        self.active = False
+        self.save()
+        return True
+
+    @property
+    def url(self):
+        return '/settings/applications/{}/'.format(self.client_id)
+
+    @property
+    def absolute_url(self):
+        return urlparse.urljoin(settings.DOMAIN, self.url)
+
+    # Properties used by Django and DRF "Links: self" field
+    @property
+    def absolute_api_v2_url(self):
```

Comment:
Will the links field refactor allow us to do away with the model method? If so I'll happily make the requisite change so that we can keep all serializer logic in one place. I support the broader goal there.